editor/src/clj/util/coll.clj

```diff
@@ -132,6 +133,41 @@
          (map (pair-fn key-fn value-fn))
          coll)))
 
+(defn partition-all-primitives
+  "Returns a lazy sequence of primitive vectors. Like core.partition-all, but
+  creates a new vector of a single primitive-type for each partition. The
+  primitive-type must be one of :int :long :float :double :byte :short :char, or
+  :boolean. Returns a stateful transducer when no collection is provided."
+  {:added "1.2"
+   :static true}
+  ([primitive-type ^long partition-length]
+   (fn [rf]
+     (let [in-progress (ArrayList. partition-length)]
+       (fn
+         ([] (rf))
+         ([result]
+          (let [result
+                (if (.isEmpty in-progress)
+                  result
+                  (let [finished (apply vector-of primitive-type (.toArray in-progress))]
+                    (.clear in-progress)
+                    (unreduced (rf result finished))))]
+            (rf result)))
+         ([result input]
+          (.add in-progress input)
+          (if (= partition-length (.size in-progress))
+            (let [finished (apply vector-of primitive-type (.toArray in-progress))]
+              (.clear in-progress)
+              (rf result finished))
+            result))))))
+  ([primitive-type ^long partition-length coll]
+   (partition-all-primitives primitive-type partition-length partition-length coll))
+  ([primitive-type ^long partition-length ^long step coll]
+   (lazy-seq
+     (when-let [in-progress (seq coll)]
+       (let [finished (apply vector-of primitive-type (take partition-length in-progress))]
+         (cons finished (partition-all-primitives primitive-type partition-length step (nthrest in-progress step))))))))
+
```

Comment:
Based on `core.partition-all`. The main difference is that we `(apply vector-of primitive-type ...)` to the `in-progress` items when we've finished a partition.

Reply:
Why do we want to use `vector-of` instead of plain vectors? I see that `Vec` stores data in primitive array, but will box the primitive on every lookup. I guess we mainly store the renderable data, and rarely access it?

Reply:
I measured it, and it doesn't appear to make a difference in this case. However, I also discovered that `reduce` is slower over a `(vector-of :double)` compared to a regular `vector` by a factor of 3-4 when I experimented with interleaving x, y, z values in a flat double vector. Again, not relevant to this case, but perhaps something to be aware of in the future.

```
From object-vector...
WARNING: Final GC required 3.340659678476196 % of runtime
Evaluation count : 359038080 in 60 samples of 5983968 calls.
             Execution time mean : 165,561063 ns
    Execution time std-deviation : 0,985169 ns
   Execution time lower quantile : 164,176282 ns ( 2,5%)
   Execution time upper quantile : 168,523537 ns (97,5%)
                   Overhead used : 1,993035 ns

Found 7 outliers in 60 samples (11,6667 %)
	low-severe	 1 (1,6667 %)
	low-mild	 3 (5,0000 %)
	high-mild	 3 (5,0000 %)
 Variance from outliers : 1,6389 % Variance is slightly inflated by outliers

From primitive-vector...
Evaluation count : 359670360 in 60 samples of 5994506 calls.
             Execution time mean : 167,817901 ns
    Execution time std-deviation : 11,173209 ns
   Execution time lower quantile : 162,723047 ns ( 2,5%)
   Execution time upper quantile : 196,444340 ns (97,5%)
                   Overhead used : 1,993035 ns

Found 4 outliers in 60 samples (6,6667 %)
	low-severe	 2 (3,3333 %)
	low-mild	 2 (3,3333 %)
 Variance from outliers : 50,0984 % Variance is severely inflated by outliers
```
Interestingly, I got the GC warning from the regular `vector` case but not the `(vector-of :double)` case. Likely just a coincidence that it triggered a GC that time, because it did not happen consistently.